hazelcast/client.py

```diff
@@ -224,3 +228,74 @@ def shutdown(self):
             self.reactor.shutdown()
             self.lifecycle.fire_lifecycle_event(LIFECYCLE_STATE_SHUTDOWN)
             self.logger.info("Client shutdown.")
+
+    def _create_address_providers(self):
+        network_config = self.config.network_config
+        address_providers = []
+
+        cloud_config = network_config.cloud_config
+        cloud_address_provider = self._init_cloud_address_provider(cloud_config)
+        if cloud_address_provider:
+            address_providers.append(cloud_address_provider)
+
+        address_providers.append(DefaultAddressProvider(network_config))
+        return address_providers
+
+    def _init_cloud_address_provider(self, cloud_config):
+        if cloud_config.enabled:
+            discovery_token = cloud_config.discovery_token
+            host, url = HazelcastCloudDiscovery.get_host_and_url(self.config.get_properties(), discovery_token)
+            return HazelcastCloudAddressProvider(host, url, self._get_connection_timeout())
+
+        cloud_token = self.properties.get(self.properties.HAZELCAST_CLOUD_DISCOVERY_TOKEN)
+        if cloud_token != "":
+            host, url = HazelcastCloudDiscovery.get_host_and_url(self.config.get_properties(), cloud_token)
+            return HazelcastCloudAddressProvider(host, url, self._get_connection_timeout())
+
+        return None
+
+    def _create_address_translator(self):
+        network_config = self.config.network_config
+        cloud_config = network_config.cloud_config
+        cloud_discovery_token = self.properties.get(self.properties.HAZELCAST_CLOUD_DISCOVERY_TOKEN)
+
+        address_list_provided = len(network_config.addresses) != 0
+        if cloud_discovery_token != "" and cloud_config.enabled:
+            raise HazelcastIllegalStateError("Ambiguous Hazelcast.cloud configuration. "
+                                             "Both property based and client configuration based settings are provided "
+                                             "for Hazelcast cloud discovery together. Use only one.")
+
+        hazelcast_cloud_enabled = cloud_discovery_token != "" or cloud_config.enabled
+        self._is_discovery_configuration_consistent(address_list_provided, hazelcast_cloud_enabled)
+
+        if hazelcast_cloud_enabled:
+            if cloud_config.enabled:
+                discovery_token = cloud_config.discovery_token
+            else:
+                discovery_token = cloud_discovery_token
+            host, url = HazelcastCloudDiscovery.get_host_and_url(self.config.get_properties(), discovery_token)
+            translator = HazelcastCloudAddressTranslator(host, url, self._get_connection_timeout())
+            if not self.config.network_config.ssl_config.enabled:
+                raise HazelcastIllegalStateError("SSL must be enabled to use Hazelcast.cloud.")
+            self.config.network_config.ssl_config.cadata = translator.cloud_discovery.get_certificate(discovery_token)
```

Comment:
Handling this over user configuration does not seem suitable to me. We will be changing the users configuration.It is not clear what will happen if users configures this. I think we should handle this internally.